### DISCUSSION TITLE
File explorer workaround until plugin system arrives

### DISCUSSION THREAD

**Discussion by cd-a** [helix-editor/helix#8314] [Q&A]
I'm really sad to see https://github.com/helix-editor/helix/pull/5768 being closed until the plugin system is ready.

While I do see the point, I was really hoping for this to be merged as it's really blocking me from working with helix properly. Not being able to browse the file structure from within the editor is a big deal for me, and I don't see the plugin system being ready anytime soon.

However, I totally understand the reasoning that was mentioned in the chat.

Until this can be done using plugins, is there some kind of workaround that people have found using wezterm or zellij magic? Like having an explorer open in another split that can actually open files in the existing helix instance in the other split?

**Comment by zetashift** [comment]
I currently use `zellij` + `yazi` https://github.com/sxyazi/yazi

```toml
[keys.normal.space.f]
d = ":sh zellij run -fc -- yazi"
```

Which gives me a file exploror I can fuzzy find through and quickly move through. Even in `neovim` I barely used the file tree.
Before `yazi` I used `xplr`.

**Reply by peteringram0** [reply]
This is a cool solution but is there a way to pipe the chosen file back to the helix window at all? Clicking `o` or `enter` when this launches a floating pane in zellij opens the selected file in the floating pane itself.

**Reply by punkbit** [reply]
Did you find how to pipe the selected file back to the hx buffers?

**Reply by ferntheplant** [reply]
I hacked a Zellij + Yazi solution together with the following scripts.

`yz-fp`
```bash
#!/bin/bash

selected_file="$1"

# KEY ASSUMPTIONS
# - editor pane was the most recently active pane
# - yazi fp is only open floating pane

# go back to editor
zellij action toggle-floating-panes

# open selected file in editor
zellij action write 27 # send escape key
zellij action write-chars ":open $selected_file"
zellij action write 13 # send enter key

# go back to floating yazi pane and close it
zellij action toggle-floating-panes
zellij action close-pane
```

`yy`
```bash
#!/bin/bash

zellij run -c -f -- yazi "$PWD"
```

I configure Yazi to use `yz-fp` as the default file opener for all files with the following snippet in my `yazi.toml`. NOTE: this effectively disables Yazi as being usable as a filepicker outside of helix+zellij
```toml
[opener]
helix = [
  { run = 'yz-fp "$0"', desc = "Use yazi as file picker within helix" },
]

[open]
rules = [
  { name = "**/*", use = "helix" },
]
```

Then within Helix I set `Ctrl+y` keybinding to invoke the `yy` script with the following snippet in my `helix/config.toml`
```toml
[keys.normal]
C-y = ":sh yy" # launch filepicker
```

My workflow is now to launch helix with no params within the root of my project and use `Ctrl+y` to launch a floating zellij pane running Yazi in the project root as well. When I open a file in this Yazi floating pane the `yz-fp` opener script will send the selected file name to the previously focused Helix pane to open said file in Helix and then close the Yazi floating pane.

**Reply by ggobbe** [reply]
This works great 🎉 thanks

**Reply by sxyazi** [reply]
Hi, I've added a simplified version on Yazi's Tips page (no need to change Yazi config!) and supports opening multiple selected files. 

Feel free to use it 😉, https://yazi-rs.github.io/docs/tips/#helix-with-zellij



https://github.com/helix-editor/helix/assets/17523360/a4dde9e0-96bf-42a4-b946-40cbee984e69

**Reply by mydoghasfleas** [reply]
Oh, this is so neat! Thank you!

**Comment by dvic** [comment]
You can have a look here: https://github.com/quantonganh/helix-wezterm https://quantonganh.com/2023/08/19/turn-helix-into-ide.md

**Reply by quantonganh** [reply]
Thank you for bringing it up.

If you are unable to access it (my home is currently experiencing an outage), you can view it [here](https://github.com/quantonganh/blog-posts/blob/main/2023/08/19/turn-helix-into-ide.md).

**Comment by omentic** [comment]
I have a fork with the file explorer patch merged in based on a commit a few months ago: https://github.com/omentic/helix/tree/staging. I've been daily driving it for a few months. There are a few stability issues related to two other patches it's running (unicode input support & moving lines directly support) but if you don't use those, it should be fine.

When Helix 23.09 releases, I plan to rebase all the patches in https://github.com/omentic/helix-patches on it, package it up, merge in a few more rejected patches, and release binaries & publish it for the AUR.

**Reply by omentic** [reply]
Also note that the file explorer is feature-complete. You won't run into the O(n^2) performance issues unless you're trying to use it in a Nix store.

**Reply by ShalokShalom** [reply]
Can I signup for a notification on that AUR package? 😄

**Reply by omentic** [reply]
@ShalokShalom Up as `helix-ext`: https://aur.archlinux.org/packages/helix-ext

**Comment by Abdillah** [comment]
I'm using [Wezterm](https://wezfurlong.org/wezterm/) + [xplr](https://github.com/sayanarijit/xplr) with [tree-view](https://github.com/sayanarijit/tree-view.xplr) plugin. With a few lua scripting.

**Comment by robrecord** [comment]
I am using a floating zellij pane with broot. 

Edit: details in the comment below.

**Reply by robrecord** [reply]
My `broot` edit file command creates a new edit pane in zellijj. Not as ideal as opening i the existing helix pane, but it works.

In `.config/broot/conf.hjson`:
```hjson
verbs: [
#...
    {
        invocation: "edit"
        keys: [
            F2
            ctrl-e
            "enter"
            right
        ]
        shortcut: "e"
        apply_to: file
        execution: "zellij edit {file} -l {line}"
        leave_broot: false
        # from_shell: true
    },
#...
]
```

**Comment by chtenb** [comment]
This threads lists several workarounds that are in use: https://github.com/helix-editor/helix/issues/2177

**Comment by TheZoq2** [comment]
I made a one-liner that does this with yazi and wezterm's split panes

```toml
[keys.normal.space]
"n" = """:sh wezterm cli split-pane sh -c "yazi --chooser-file=/dev/stdout | while read -r; do wezterm cli send-text --pane-id ${WEZTERM_PANE} --no-paste \\":o \\${REPLY}\\"; done" > /dev/null"""

```

**Reply by acramsay** [reply]
Thanks for this awesome one liner!

FYI for those like me that just discovered this, you can send a "return" character as well instead of mashing the Enter key twice. Here's my little tweak.

```sh
[keys.normal.space]
"o" = """:sh wezterm cli split-pane sh -c "yazi --chooser-file=/dev/stdout | while read -r; do wezterm cli send-text --pane-id ${WEZTERM_PANE} --no-paste \\":o \\${REPLY}\r\\"; done" > /dev/null"""
```

**Comment by lukeflo** [comment]
Just for the sake of completeness, I show my solution using `foot` terminal, `sway` window-manager and `fff` or `yazi` file manager on Void Linux (not the most common combination, but maybe some day someone is searching for it... :grin: ). Thus, no terminal multiplexer, wezterm etc. needed (demo gif using `fff`):

![recorded](https://github.com/user-attachments/assets/40fe02ef-ecea-4b2e-928e-0aae4fe84435)

For Helix PR [#11164](https://github.com/helix-editor/helix/pull/11164) is very helpful too.

## sway config

Inside my sway config I added following line to make floating terminal windows possible:

```config
for_window [app_id = "floating_foot"] floating enable
```

## script

Furthermore, I need a small wrapper script which opens a new `foot` window with the correct `app_id` to make it float. The script is also necessary since it combines the `fff` or `yazi` run and the following printing of the file path to `stdout`:

```bash
#!/usr/bin/env bash

# Send picker as $1, save selected filepath in cache file
# and print filepath to stdout using cat
# -a `floating_foot` sets app_id for floating term window
# -p option to fff saves selected filepath in tmp-file instead of opening
case "$1" in
fff)
    foot -a 'floating_foot' zsh -c 'fff -p' && cat "/home/lukeflo/.config/cache/fff/opened_file"
    ;;
yazi)
    tmp="$(mktemp -t yazi-picker.XXXXXX)"
    foot -a 'floating_foot' yazi --chooser-file "$tmp"
    cat -- "$tmp"
    rm -f -- "$tmp"
    ;;
*)
    echo "Unknown picker"
    ;;
esac
```

## Helix shortcut

Finally, I just added the script call as shortcut to my `config.toml`:

```toml
[keys.normal.space]
"o" = [ ":open %sh{ sway-picker fff }" ]
```

To use `yazi`, just replace `fff` with `yazi`.

Now I can simply press `space` `o` in normal mode. A floating window running `fff`/`yazi` pops up and I can select a file which will be opened in my current Helix buffer...

**Comment by andsilmar** [comment]
Does anyone know if the approach described here for Git integration (https://github.com/helix-editor/helix/wiki/Recipes#git-integration) works with a file tree? I've tried a couple of options but haven't managed to get it working.

**Reply by quantonganh** [reply]
What do you mean by "work with a file tree"?

**Reply by andsilmar** [reply]
You're right, I didn't explain myself very well. I want to use Yazi or similar software to explore the project's files and perform actions like creating, editing, and bulk renaming files. Does that make sense?

**Reply by ggobbe** [reply]
I use `yazi` and `zellij` to have a file explorer in helix. I can create delete files with yazi [common file operations](https://yazi-rs.github.io/docs/quick-start/#file-operations), and hit ENTER to open a file in helix.

My only downside is that it always starts from the directory in which I ran `hx .` where I'd prefer it to start from the directory of the currently opened buffer.

Here is a summary of my config, mostly taken from a [comment](https://github.com/helix-editor/helix/discussions/8314#discussioncomment-8921134) above.

**~/.config/helix/config.toml**
```toml
[keys.normal.space]
"." = ":sh yy"

[keys.normal]
C-y = ":sh yy"
```

**~/.local/bin/yy**
```bash
#!/bin/bash

zellij run -c -f --x 10% --y 10% --width 80% --height 80% -- yazi "$PWD"
```

**~/.config/yazi/yazi.toml**
```toml
[opener]
text = [
        { run = 'hx "$@"', block = true },
]
preview = [
        { run = 'open "$@"'}
]
helix = [
  { run = 'yz-fp "$0"', desc = "Use yazi as file picker within helix" },
]

[manager]
sort_dir_first = true
ratio = [2, 2, 5]

[open]
rules = [
  { name = "**/*", use = "helix" },
]
```

**~/.local/bin/yz-fp**
```bash
#!/bin/bash

selected_file="$1"

# KEY ASSUMPTIONS
# - editor pane was the most recently active pane
# - yazi fp is only open floating pane

# go back to editor
zellij action toggle-floating-panes

# open selected file in editor
zellij action write 27 # send escape key
zellij action write-chars ":open $selected_file"
zellij action write 13 # send enter key

# go back to floating yazi pane and close it
zellij action toggle-floating-panes
zellij action close-pane
```

**Preview**
<img width="2053" alt="image" src="https://github.com/user-attachments/assets/a3a89150-c9a3-4568-afd1-2028d8983544" />

**Reply by andsilmar** [reply]
Thank you!

**Reply by ferntheplant** [reply]
@ggobbe [soon you'll be able to make it start from the current buffer](https://github.com/helix-editor/helix/discussions/11899#discussioncomment-11256813)!

**Reply by quantonganh** [reply]
@andsilmar Are you aware of these PRs?
- [Add file explorer](https://github.com/helix-editor/helix/pull/11285)
- [Create, Delete, Copy, Move, Rename files in the new File Explorer](https://github.com/helix-editor/helix/pull/12902)

In case you still want to use yazi, IMO, it would be better to open it in a split/floating pane (if your terminal supports it) rather than using `:insert-output` and `:redraw`: https://www.reddit.com/r/HelixEditor/comments/1bgsauh/instruction_how_to_setup_file_tree_in_helix_using/

**Reply by andsilmar** [reply]
Thank you very much, @quantonganh. Your recommendations work for me. I'm actually using Yazi because I don’t have a native file explorer inside Helix—this would be ideal. For now, I’ll follow your recommendation.

**Comment by godalming123** [comment]
Here's [my current workarond](https://github.com/godalming123/helixdots/blob/ae00dca6f91af8fd4b97c92ad136edca6674e752/config.toml#L28), which based on [this](https://gist.github.com/lukepighetti/8e4a13db5bdcd68a7d83eee19051ab14). I also had an [old workaround](https://github.com/godalming123/helixdots/blob/11b10727257ea984670d704f2d88a29a1ee97a74/config.toml#L51) which only used the coreutils, but was quite buggy.

**Reply by kumawatdarshan** [reply]
Yazi now natively supports plugging into helix. It's very seamless.

**Comment by Abdillah** [comment]
Helix has added [a file explorer](https://github.com/helix-editor/helix/pull/11285), but I haven't tried because it is not yet released.
Anyone tried yet?

**Reply by Android789515** [reply]
It's just a harder to use fuzzy finder. You have to navigate from base project dir every time you open the explorer, no matter where your buffer is located.

**Reply by zetashift** [reply]
@Android789515 there is `file_browser_in_current_buffer_directory` for opening the browser in the directory of the current buffer.

I still use `yazi` though, it does a lot more.

**Comment by DoctorRainer** [comment]
Yazi works great right now, I use this binding and nothing else for it to work:

```toml
y = [
  ':sh rm -f /tmp/yazi-path',
  ':insert-output yazi %{buffer_name} --chooser-file=/tmp/yazi-path',
  ":open %sh{cat /tmp/yazi-path}",
  ":redraw",
  ":reload-all",
  ':set mouse false',
  ':set mouse true'
]
```

**Comment by luccahuguet** [comment]
after developing https://github.com/luccahuguet/yazelix, there is nothing i miss from an integrated file explorer
<img width="1451" height="718" alt="image" src="https://github.com/user-attachments/assets/8a16c370-60ec-447d-bce7-c3011d29ada2" />
